server/src/main/java/io/pravega/schemaregistry/server/rest/RestServer.java

```diff
@@ -0,0 +1,103 @@
+import org.glassfish.jersey.server.ResourceConfig;
+import org.glassfish.jersey.server.ServerProperties;
+
+/**
+ * Netty REST server implementation.
+ */
+@Slf4j
+public class RestServer extends AbstractIdleService {
+    private final String objectId;
+    private final ServiceConfig restServerConfig;
+    private final URI baseUri;
+    private final ResourceConfig resourceConfig;
+    private HttpServer httpServer;
+
+    public RestServer(SchemaRegistryService registryService, ServiceConfig restServerConfig) {
+        this.objectId = "RestServer";
+        this.restServerConfig = restServerConfig;
+        this.baseUri = UriBuilder.fromUri("http://" + restServerConfig.getHost()).port(restServerConfig.getPort()).build();
+
+        final Set<Object> resourceObjs = new HashSet<>();
+        resourceObjs.add(new PingImpl());
+        resourceObjs.add(new GroupResourceImpl(registryService, restServerConfig, executor()));
+        resourceObjs.add(new SchemaResourceImpl(registryService, restServerConfig, executor()));
+
+        final RegistryApplication application = new RegistryApplication(resourceObjs);
+        this.resourceConfig = ResourceConfig.forApplication(application);
+        this.resourceConfig.property(ServerProperties.BV_SEND_ERROR_IN_RESPONSE, true);
+
+        // Register the custom JSON parser.
+        this.resourceConfig.register(new CustomObjectMapperProvider());
+    }
+
+    /**
+     * Start REST service.
+     */
+    @Override
+    protected void startUp() {
+        long traceId = LoggerHelpers.traceEnterWithContext(log, this.objectId, "startUp");
+        try {
+            log.info("Starting REST server listening on port: {}", this.restServerConfig.getPort());
+            if (restServerConfig.isTlsEnabled()) {
+                SSLContextConfigurator contextConfigurator = new SSLContextConfigurator();
+                contextConfigurator.setKeyStoreFile(restServerConfig.getTlsKeyFilePath());
+                contextConfigurator.setKeyStorePass(JKSHelper.loadPasswordFrom(restServerConfig.getTlsKeyPasswordFilePath()));
+                httpServer = GrizzlyHttpServerFactory.createHttpServer(baseUri, resourceConfig, true,
+                        new SSLEngineConfigurator(contextConfigurator, false, false, false));
+            } else {
+                httpServer = GrizzlyHttpServerFactory.createHttpServer(baseUri, resourceConfig, true);
+            }
+
+        } finally {
+            LoggerHelpers.traceLeave(log, this.objectId, "startUp", traceId);
+        }
+    }
+
+    /**
+     * Gracefully stop REST service.
+     */
+    @Override
+    protected void shutDown() throws Exception {
+        long traceId = LoggerHelpers.traceEnterWithContext(log, this.objectId, "shutDown");
+        try {
+            log.info("Stopping REST server listening on port: {}", this.restServerConfig.getPort());
+            final GrizzlyFuture<HttpServer> shutdown = httpServer.shutdown(30, TimeUnit.SECONDS);
```

Comment:
Curious, what happens when the 30 seconds limit is crossed. Does the resources still get cleaned up in the background? Should we have a longer timeout here?

Reply:
It force shutsdown. 
Basically we attempt graceful shutdown but force it after 30 seconds lest the service gets stuck in shutdown phase which can prevent a requested restart.